File: app/src/main/java/net/xaethos/trackernotifier/fragments/StoryCommentsFragment.java
package net.xaethos.trackernotifier.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import net.xaethos.trackernotifier.R;
import net.xaethos.trackernotifier.adapters.CommentsDataSource;
import net.xaethos.trackernotifier.adapters.ResourceAdapter;
import net.xaethos.trackernotifier.api.TrackerClient;
import net.xaethos.trackernotifier.models.Comment;
import net.xaethos.trackernotifier.models.Story;

import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;

public class StoryCommentsFragment extends BaseResourceFragment<Story> {

    TrackerClient mApiClient;
    View mEmptyView;
    ResourceAdapter<Comment, CommentsDataSource> mAdapter;

    long mProjectId;
    Subscription mContentSubscription;

    public static StoryCommentsFragment newInstance(long projectId) {
        Bundle args = new Bundle(1);
        args.putLong("projectId", projectId);

        StoryCommentsFragment fragment = new StoryCommentsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mProjectId = getArguments().getLong("projectId");
        mApiClient = TrackerClient.getInstance();
        mAdapter = CommentsDataSource.createAdapter();
    }

    @Nullable
    @Override
    public View onCreateView(
            LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_notifications, container, false);

        RecyclerView recyclerView = (RecyclerView) root.findViewById(android.R.id.list);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(mAdapter);

        mContentSubscription =
                observeResource().flatMap((story) -> mApiClient.comments.get(mProjectId, story.id))
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe((comments -> {
                            mAdapter.getDataSource().setComments(comments);
                        }));

        return root;
    }

    @Override
    public void onDestroyView() {
        mContentSubscription.unsubscribe();
        super.onDestroyView();
    }

}

File: app/src/main/java/net/xaethos/trackernotifier/fragments/StoryCommentsFragment.kt
package net.xaethos.trackernotifier.fragments

import android.content.Context
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import net.xaethos.trackernotifier.R
import net.xaethos.trackernotifier.adapters.CommentViewHolder
import net.xaethos.trackernotifier.adapters.ResourceAdapter
import net.xaethos.trackernotifier.adapters.SimpleDividerDecorator
import net.xaethos.trackernotifier.api.TrackerClient
import net.xaethos.trackernotifier.models.Story
import rx.Observable
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers

class StoryCommentsFragment : BaseAdapterFragment() {
    protected var storyObservable: Observable<Story>? = null
        private set


    private val apiClient = TrackerClient.instance

    override protected val adapter = ResourceAdapter(R.layout.item_comment) { inflater, viewType, parent ->
        CommentViewHolder(inflater.inflate(viewType, parent, false))
    }

    private var projectId: Long = 0
    private var contentSubscription: Subscription? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        storyObservable = (context as? BaseResourceFragment.ResourceSource<Story>)?.resourceObservable
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        projectId = arguments.getLong("projectId")
    }

    override fun onCreateView(
            inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?)
            = inflater.inflate(R.layout.fragment_list, container, false)

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recyclerView?.layoutManager = LinearLayoutManager(context)
        recyclerView?.addItemDecoration(SimpleDividerDecorator(context))
        recyclerView?.adapter = adapter

        contentSubscription = subscribeComments()
    }

    override fun onDestroyView() {
        contentSubscription?.unsubscribe()
        super.onDestroyView()
    }

    private fun subscribeComments(): Subscription {
        setEmptyText(R.string.msg_fetching_content, 0)
        return (storyObservable ?: Observable.empty<Story>())
                .take(1)
                .flatMap { story -> apiClient.comments.get(projectId, story.id, ":default,person(name,username,initials)") }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ comments -> adapter.setResources(comments) },
                        { error -> setEmptyText(getString(R.string.headline_error), error.message) },
                        { setEmptyText(R.string.headline_no_comments, 0) })
    }

    companion object {
        fun newInstance(projectId: Long): StoryCommentsFragment {
            val args = Bundle(1)
            args.putLong("projectId", projectId)

            val fragment = StoryCommentsFragment()
            fragment.arguments = args
            return fragment
        }
    }

}
